core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -309,36 +311,47 @@ private UpdateResultHolder executeUpdateRequest(UpdateRequest updateRequest, Ind
-                            throw new IllegalStateException("Unexpected request operation type on replica: "
-                                + docWriteRequest.opType().getLowercase());
+                    if (item.getPrimaryResponse().isFailed()) {
+                        // execution on primary resulted in a failure
+                        // if primary execution generated a sequence no, execute a noop on the replica engine to record it in the translog
+                        final BulkItemResponse.Failure failure = item.getPrimaryResponse().getFailure();
+                        operationResult = failure.getSeqNo() != SequenceNumbersService.UNASSIGNED_SEQ_NO
+                                ? executeNoOpRequestOnReplica(failure, docWriteRequest, replica)
+                                : null;
+                    } else {
+                        final DocWriteResponse primaryResponse = item.getPrimaryResponse().getResponse();
+                        switch (docWriteRequest.opType()) {
+                            case CREATE:
+                            case INDEX:
+                                operationResult = executeIndexRequestOnReplica(primaryResponse, (IndexRequest) docWriteRequest, replica);
+                                break;
+                            case DELETE:
+                                operationResult = executeDeleteRequestOnReplica(primaryResponse, (DeleteRequest) docWriteRequest, replica);
+                                break;
+                            default:
+                                throw new IllegalStateException("Unexpected request operation type on replica: "
+                                        + docWriteRequest.opType().getLowercase());
+                        }
+                        assert operationResult != null : "operation result must never be null when primary response has no failure";
                     }
-                    if (operationResult.hasFailure()) {
-                        // check if any transient write operation failures should be bubbled up
-                        Exception failure = operationResult.getFailure();
-                        assert failure instanceof VersionConflictEngineException
-                            || failure instanceof MapperParsingException
-                            : "expected any one of [version conflict, mapper parsing, engine closed, index shard closed]" +
-                            " failures. got " + failure;
-                        if (!TransportActions.isShardNotAvailableException(failure)) {
-                            throw failure;
+                    if (operationResult != null) {
+                        if (operationResult.hasFailure()) {
+                            // check if any transient write operation failures should be bubbled up
+                            Exception failure = operationResult.getFailure();
+                            assert failure instanceof VersionConflictEngineException
```

Comment:
FYI we don't throw version conflicts on replicas any more...  (not related to your change)